### DISCUSSION TITLE
Failed SMTP authentication, OAuth token doesn't seem to be sent to server

### DISCUSSION THREAD

**Discussion by CKCordes** [jstedfast/MailKit#1658] [Q&A] [answered]
I am using an oauth2 token to authenticate a user for a mail client that i am wokring on. I can successfully connect and authenticate an IMAPClient, however when i try to do the same for the SMTPClient, i am met with the exception 

S: 501-5.5.2 Cannot Decode response. Please visit
S: 501 5.5.2  https://support.google.com/a/answer/3221692 for more information.

When i check the protocol logger, i seems that the oauth token is not sent to the server, however i am new to all this so i might be wrong! :)
The full logs look like this:

```
Connected to smtps://smtp.gmail.com:465/
S: 220 smtp.gmail.com --long string that ends with gsmtp, dont know if private--
C: EHLO --PCNAME--
S: 250-smtp.gmail.com at your service,--IP--
S: 250-SIZE 35882577
S: 250-8BITMIME
S: 250-AUTH LOGIN PLAIN XOAUTH2 PLAIN-CLIENTTOKEN OAUTHBEARER XOAUTH
S: 250-ENHANCEDSTATUSCODES
S: 250-PIPELINING
S: 250-CHUNKING
S: 250 SMTPUTF8
C: AUTH XOAUTH2 
S: 334 
C: 
S: 501-5.5.2 Cannot Decode response. Please visit
S: 501 5.5.2  https://support.google.com/a/answer/3221692 for more information. --long string that ends with gsmtp, dont know if private--
```
The reason i say i don't think the token isn't sent to the server is because in the first client response after EHLO, there isn't any base64 string or anything, which i have seen in other examples of logs...

I attempt to connect to the SMTPClient in the following way:

```
var initiatedSmtpClient = new SmtpClient(new ProtocolLogger("smtp"));
initiatedSmtpClient.Connect("smtp.gmail.com", 465, true);
initiatedSmtpClient.Authenticate(Credentials.oAuth2);
```

Where the Credentials class is a static class containing the SaslMechanismOAuth2.  I use pretty much the same method for the IMAPClient, which works out...

I hope somebody can help me out, i am completely out of ideas. Thanks alot!

**Comment by jstedfast** [comment, marked as answer]
If Credentials.oauth2 is a SaslMechanismOAuth2, you need to reset it after authenticating before reuse.

**Reply by CKCordes** [reply]
That's exactly it, thanks alot! By the way great job on the library, it is really well done!
